### PR TITLE
FT connector:  add prefix to results

### PR DESCRIPTION
**Pull request by karim-en**
A prefix was added to results structures to avoid collisions on decoding the proof

### CODE REVIEW

File: bridge-common/src/result_types.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    255, 227, 34, 202, 249, 14, 225, 115, 248, 131, 171,
+];
+pub const RESULT_PREFIX_LOCK: ResultPrefix = [
+    10, 158, 184, 119, 69, 133, 121, 219, 206, 131, 234, 87, 213, 86, 190, 80, 209, 195, 22, 11,
+    181, 241, 113, 159, 177, 114, 189, 51, 0, 172, 134, 35,
+];
+pub const RESULT_PREFIX_METADATA: ResultPrefix = [
+    179, 21, 212, 214, 232, 242, 53, 245, 250, 187, 11, 26, 15, 17, 133, 7, 246, 200, 84, 47, 174,
+    142, 26, 149, 102, 171, 230, 7, 98, 4, 124, 22,
+];
```
Comment (sept-en) [inline bridge-common/src/result_types.rs:18]: Let's add an explanation so we don't need to check the tests to figure this out
```suggestion
/// This prefix value is computed as: `keccak256(b"ResultType.Withdraw").as_slice()`
pub const RESULT_PREFIX_WITHDRAW: ResultPrefix = [
    246, 181, 82, 1, 204, 142, 13, 135, 141, 218, 111, 185, 98, 41, 84, 186, 141, 37, 252, 127, 56,
    255, 227, 34, 202, 249, 14, 225, 115, 248, 131, 171,
];

/// This prefix value is computed as: `keccak256(b"ResultType.Lock").as_slice()`
pub const RESULT_PREFIX_LOCK: ResultPrefix = [
    10, 158, 184, 119, 69, 133, 121, 219, 206, 131, 234, 87, 213, 86, 190, 80, 209, 195, 22, 11,
    181, 241, 113, 159, 177, 114, 189, 51, 0, 172, 134, 35,
];

/// This prefix value is computed as: `keccak256(b"ResultType.Metadata").as_slice()`
pub const RESULT_PREFIX_METADATA: ResultPrefix = [
    179, 21, 212, 214, 232, 242, 53, 245, 250, 187, 11, 26, 15, 17, 133, 7, 246, 200, 84, 47, 174,
    142, 26, 149, 102, 171, 230, 7, 98, 4, 124, 22,
];
```

File: bridge-common/src/result_types.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    fn default() -> Self {
+        Self {
+            #[cfg(feature = "result_with_prefix")]
+            prefix: RESULT_PREFIX_WITHDRAW,
+            amount: Default::default(),
+            token: Default::default(),
+            recipient: Default::default(),
+        }
+    }
+}
```
Comment (sept-en) [inline bridge-common/src/result_types.rs:39]: It's easy to rely on that everyone will use the `Default` trait. What we actually need is to implement some kind of construct so it's always enforced to use that specific prefix value for each of the types.

File: bridge-common/src/result_types.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    println!(
+        "RESULT_PREFIX_WITHDRAW: {}",
+        hex::encode(RESULT_PREFIX_WITHDRAW)
+    );
+    println!("RESULT_PREFIX_LOCK: {}", hex::encode(RESULT_PREFIX_LOCK));
+    println!(
+        "RESULT_PREFIX_METADATA: {}",
+        hex::encode(RESULT_PREFIX_METADATA)
+    );
```
Comment (sept-en) [inline bridge-common/src/result_types.rs:110]: I don't think this is still needed, right? I think what can be done here is to compare hex-representation with bytes encoding (so it's visible on Ethereum impl's side that it's the same value)
```suggestion
```

**Review by sept-en — CHANGES_REQUESTED** [review]

File: bridge-common/src/result_types.rs (hunk 3 above)
Comment (karim-en) [reply bridge-common/src/result_types.rs]: Fixed 9b681b9059f63682bab63dfc29dfe768d7997916

File: bridge-common/src/result_types.rs (hunk 2 above)
Comment (karim-en) [reply bridge-common/src/result_types.rs]: Fixed 9b681b9059f63682bab63dfc29dfe768d7997916

File: token-locker/src/lib.rs [hunk 4]
```diff
@@ -125,28 +135,31 @@ impl Contract {
...
-            decimals: metadata.decimals,
-            block_height: env::block_height(),
-        }
+    ) -> result_types::Metadata {
+        result_types::Metadata::new(
+            env::predecessor_account_id().into(),
+            metadata.name,
+            metadata.symbol,
+            metadata.decimals,
+            env::block_height()
```
Comment (sept-en) [inline token-locker/src/lib.rs:161]:
```suggestion
            env::block_height(),
```

File: token-locker/src/lib.rs [hunk 5]
```diff
@@ -215,37 +222,30 @@ impl Contract {
...
-            .with_static_gas(FT_TRANSFER_CALL_GAS)
-            .ft_transfer_call(
-                    target.try_into().unwrap(),
-                    amount.into(),
-                    None,
-                    message,
-                ),
+                .with_attached_deposit(1)
+                .with_static_gas(FT_TRANSFER_CALL_GAS)
+                .ft_transfer_call(target.try_into().unwrap(), amount.into(), None, message),
```
Comment (sept-en) [inline token-locker/src/lib.rs:241]: It's already `AccountId`
```suggestion
                .ft_transfer_call(target.unwrap(), amount.into(), None, message),
```

File: token-locker/src/lib.rs [hunk 6]
```diff
@@ -215,37 +222,30 @@ impl Contract {
...
             Some(message) => ext_token::ext(token.try_into().unwrap())
-            .with_attached_deposit(1)
-            .with_static_gas(FT_TRANSFER_CALL_GAS)
-            .ft_transfer_call(
-                    target.try_into().unwrap(),
-                    amount.into(),
-                    None,
-                    message,
-                ),
+                .with_attached_deposit(1)
```
Comment (sept-en) [inline token-locker/src/lib.rs:239]: Let's use these constants now
```suggestion
                .with_attached_deposit(near_sdk::ONE_YOCTO)
```

File: token-locker/src/lib.rs [hunk 7]
```diff
@@ -215,37 +222,30 @@ impl Contract {
...
+                .ft_transfer_call(target.try_into().unwrap(), amount.into(), None, message),
             None => ext_token::ext(token.try_into().unwrap())
-            .with_attached_deposit(1)
-            .with_static_gas(FT_TRANSFER_GAS)
-            .ft_transfer(
-                target.into(),
-                amount.into(),
-                None,
-            ),
+                .with_attached_deposit(1)
```
Comment (sept-en) [inline token-locker/src/lib.rs:243]:
```suggestion
                .with_attached_deposit(near_sdk::ONE_YOCTO)
```

File: token-locker/src/lib.rs [hunk 8]
```diff
@@ -215,37 +222,30 @@ impl Contract {
...
+                .with_static_gas(FT_TRANSFER_GAS)
+                .ft_transfer(target.into(), amount.into(), None),
```
Comment (sept-en) [inline token-locker/src/lib.rs:245]:
```suggestion
                .ft_transfer(target, amount.into(), None),
```

**Review by sept-en — CHANGES_REQUESTED** [review]

**Review by sept-en — APPROVED** [review]

**Pull request merged**